File: dbflow-tests/src/androidTest/java/com/raizlabs/android/dbflow/test/provider/TestContentProvider.java
package com.raizlabs.android.dbflow.test.provider;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;

import com.raizlabs.android.dbflow.annotation.provider.ContentProvider;
import com.raizlabs.android.dbflow.annotation.provider.ContentUri;
import com.raizlabs.android.dbflow.annotation.provider.Notify;
import com.raizlabs.android.dbflow.annotation.provider.TableEndpoint;
import com.raizlabs.android.dbflow.sql.SqlUtils;

/**
 * Description:
 */
@ContentProvider(authority = TestContentProvider.AUTHORITY,
        database = ContentDatabase.class,
        baseContentUri = TestContentProvider.BASE_CONTENT_URI)
public class TestContentProvider {

    public static final String AUTHORITY = "com.raizlabs.android.dbflow.test.provider";

    public static final String BASE_CONTENT_URI = "content://";

    private static Uri buildUri(String... paths) {
        Uri.Builder builder = Uri.parse(BASE_CONTENT_URI + AUTHORITY).buildUpon();
        for (String path : paths) {
            builder.appendPath(path);
        }
        return builder.build();
    }

    @TableEndpoint(name = ContentProviderModel.ENDPOINT, contentProvider = ContentDatabase.class)
    public static class ContentProviderModel {

        public static final String ENDPOINT = "ContentProviderModel";

        @ContentUri(path = ContentProviderModel.ENDPOINT,
                type = ContentUri.ContentType.VND_MULTIPLE + ENDPOINT)
        public static Uri CONTENT_URI = buildUri(ENDPOINT);

        @ContentUri(path = ContentProviderModel.ENDPOINT + "/#",
                type = ContentUri.ContentType.VND_SINGLE + ENDPOINT,
                segments = {@ContentUri.PathSegment(segment = 1, column = "id")})
        public static Uri withId(long id) {
            return buildUri(String.valueOf(id));
        }

        @Notify(method = Notify.Method.INSERT,
                paths = ContentProviderModel.ENDPOINT + "/#")
        public static Uri[] onInsert(ContentValues contentValues) {
            final long id = contentValues.getAsLong("id");
            return new Uri[]{
                    withId(id)
            };
        }

    }

    @TableEndpoint(name = NoteModel.ENDPOINT, contentProvider = ContentDatabase.class)
    public static class NoteModel {

        public static final String ENDPOINT = "NoteModel";

        @ContentUri(path = ENDPOINT,
                type = ContentUri.ContentType.VND_MULTIPLE + ENDPOINT)
        public static Uri CONTENT_URI = buildUri(ENDPOINT);

        @ContentUri(path = ENDPOINT + "/#",
                type = ContentUri.ContentType.VND_MULTIPLE + ENDPOINT,
                segments = {@ContentUri.PathSegment(column = "id", segment = 1)})
        public static Uri withId(long id) {
            return buildUri(ENDPOINT, String.valueOf(id));
        }

        @ContentUri(path = ENDPOINT + "/#/#",
                type = ContentUri.ContentType.VND_SINGLE + ContentProviderModel.ENDPOINT,
                segments = {@ContentUri.PathSegment(column = "id", segment = 2)})
        public static Uri fromList(long id) {
            return buildUri(ENDPOINT, "fromList", String.valueOf(id));
        }

        @ContentUri(path = ENDPOINT + "/#/#",
                type = ContentUri.ContentType.VND_SINGLE + ContentProviderModel.ENDPOINT,
                segments = {@ContentUri.PathSegment(column = "id", segment = 1),
                        @ContentUri.PathSegment(column = "isOpen", segment = 2)})
        public static Uri withOpenId(long id, boolean isOpen) {
            return buildUri(ENDPOINT, String.valueOf(id), String.valueOf(isOpen));
        }

        @Notify(method = Notify.Method.INSERT, paths = ENDPOINT)
        public static Uri[] onInsert(ContentValues contentValues) {
            final long listId = contentValues.getAsLong(SqlUtils.getContentValuesKey(contentValues, "providerModel"));
            return new Uri[]{
                    ContentProviderModel.withId(listId), fromList(listId),
            };
        }

        @Notify(method = Notify.Method.INSERT, paths = ENDPOINT)
        public static Uri onInsert2(ContentValues contentValues) {
            final long listId = contentValues.getAsLong(SqlUtils.getContentValuesKey(contentValues, "providerModel"));
            return fromList(listId);
        }

        @Notify(method = Notify.Method.UPDATE, paths = ENDPOINT + "/#")
        public static Uri[] onUpdate(Context context, Uri uri) {
            final long noteId = Long.valueOf(uri.getPathSegments().get(1));
            Cursor c = context.getContentResolver().query(uri, new String[]{
                    "noteModel",
            }, null, null, null);
            c.moveToFirst();
            final long listId = c.getLong(c.getColumnIndex("providerModel"));
            c.close();

            return new Uri[]{
                    withId(noteId), fromList(listId), ContentProviderModel.withId(listId),
            };
        }

        @Notify(method = Notify.Method.DELETE, paths = ENDPOINT + "/#")
        public static Uri[] onDelete(Context context, Uri uri) {
            final long noteId = Long.valueOf(uri.getPathSegments().get(1));
            Cursor c = context.getContentResolver().query(uri, null, null, null, null);
            c.moveToFirst();
            final long listId = c.getLong(c.getColumnIndex("providerModel"));
            c.close();

            return new Uri[]{
                    withId(noteId), fromList(listId), ContentProviderModel.withId(listId),
            };
        }
    }

    @TableEndpoint(name = TestSyncableModel.ENDPOINT, contentProvider = ContentDatabase.class)
    public static class TestSyncableModel {

        public static final String ENDPOINT = "TestSyncableModel";

        @ContentUri(path = ENDPOINT,
                type = ContentUri.ContentType.VND_MULTIPLE + ENDPOINT)
        public static Uri CONTENT_URI = buildUri(ENDPOINT);
    }
}

File: dbflow-tests/src/test/java/com/raizlabs/android/dbflow/provider/TestContentProvider.kt
package com.raizlabs.android.dbflow.provider

import android.content.ContentValues
import android.content.Context
import android.net.Uri
import com.raizlabs.android.dbflow.annotation.provider.ContentProvider
import com.raizlabs.android.dbflow.annotation.provider.ContentUri
import com.raizlabs.android.dbflow.annotation.provider.Notify
import com.raizlabs.android.dbflow.annotation.provider.TableEndpoint
import com.raizlabs.android.dbflow.sql.SqlUtils

/**
 * Description:
 */
@ContentProvider(authority = TestContentProvider.AUTHORITY, database = ContentDatabase::class,
    baseContentUri = TestContentProvider.BASE_CONTENT_URI)
object TestContentProvider {

    const val AUTHORITY = "com.raizlabs.android.dbflow.test.provider"

    const val BASE_CONTENT_URI = "content://"

    private fun buildUri(vararg paths: String): Uri {
        val builder = Uri.parse(BASE_CONTENT_URI + AUTHORITY).buildUpon()
        for (path in paths) {
            builder.appendPath(path)
        }
        return builder.build()
    }

    @TableEndpoint(name = ContentProviderModel.ENDPOINT, contentProvider = ContentDatabase::class)
    object ContentProviderModel {

        const val ENDPOINT = "ContentProviderModel"

        @JvmStatic
        @ContentUri(path = ContentProviderModel.ENDPOINT,
            type = ContentUri.ContentType.VND_MULTIPLE + ENDPOINT)
        var CONTENT_URI = buildUri(ENDPOINT)

        @JvmStatic
        @ContentUri(path = ContentProviderModel.ENDPOINT + "/#",
            type = ContentUri.ContentType.VND_SINGLE + ENDPOINT,
            segments = arrayOf(ContentUri.PathSegment(segment = 1, column = "id")))
        fun withId(id: Long): Uri {
            return buildUri(id.toString())
        }

        @JvmStatic
        @Notify(method = Notify.Method.INSERT, paths = arrayOf(ContentProviderModel.ENDPOINT + "/#"))
        fun onInsert(contentValues: ContentValues): Array<Uri> {
            val id = contentValues.getAsLong("id")!!
            return arrayOf(withId(id))
        }

    }

    @TableEndpoint(name = NoteModel.ENDPOINT, contentProvider = ContentDatabase::class)
    object NoteModel {

        const val ENDPOINT = "NoteModel"

        @ContentUri(path = ENDPOINT, type = ContentUri.ContentType.VND_MULTIPLE + ENDPOINT)
        var CONTENT_URI = buildUri(ENDPOINT)

        @JvmStatic
        @ContentUri(path = ENDPOINT + "/#", type = ContentUri.ContentType.VND_MULTIPLE + ENDPOINT,
            segments = arrayOf(ContentUri.PathSegment(column = "id", segment = 1)))
        fun withId(id: Long): Uri {
            return buildUri(ENDPOINT, id.toString())
        }

        @JvmStatic
        @ContentUri(path = ENDPOINT + "/#/#",
            type = ContentUri.ContentType.VND_SINGLE + ContentProviderModel.ENDPOINT,
            segments = arrayOf(ContentUri.PathSegment(column = "id", segment = 2)))
        fun fromList(id: Long): Uri {
            return buildUri(ENDPOINT, "fromList", id.toString())
        }

        @JvmStatic
        @ContentUri(path = ENDPOINT + "/#/#",
            type = ContentUri.ContentType.VND_SINGLE + ContentProviderModel.ENDPOINT,
            segments = arrayOf(ContentUri.PathSegment(column = "id", segment = 1),
                ContentUri.PathSegment(column = "isOpen", segment = 2)))
        fun withOpenId(id: Long, isOpen: Boolean): Uri {
            return buildUri(ENDPOINT, id.toString(), isOpen.toString())
        }

        @JvmStatic
        @Notify(method = Notify.Method.INSERT, paths = arrayOf(ENDPOINT))
        fun onInsert(contentValues: ContentValues): Array<Uri> {
            val listId = contentValues.getAsLong(SqlUtils.getContentValuesKey(contentValues, "providerModel"))!!
            return arrayOf(ContentProviderModel.withId(listId), fromList(listId))
        }

        @JvmStatic
        @Notify(method = Notify.Method.INSERT, paths = arrayOf(ENDPOINT))
        fun onInsert2(contentValues: ContentValues): Uri {
            val listId = contentValues.getAsLong(SqlUtils.getContentValuesKey(contentValues, "providerModel"))!!
            return fromList(listId)
        }

        @JvmStatic
        @Notify(method = Notify.Method.UPDATE, paths = arrayOf(ENDPOINT + "/#"))
        fun onUpdate(context: Context, uri: Uri): Array<Uri> {
            val noteId = java.lang.Long.valueOf(uri.pathSegments[1])!!
            val c = context.contentResolver.query(uri, arrayOf("noteModel"), null, null, null)
            c!!.moveToFirst()
            val listId = c.getLong(c.getColumnIndex("providerModel"))
            c.close()

            return arrayOf(withId(noteId), fromList(listId), ContentProviderModel.withId(listId))
        }

        @JvmStatic
        @Notify(method = Notify.Method.DELETE, paths = arrayOf(ENDPOINT + "/#"))
        fun onDelete(context: Context, uri: Uri): Array<Uri> {
            val noteId = java.lang.Long.valueOf(uri.pathSegments[1])!!
            val c = context.contentResolver.query(uri, null, null, null, null)
            c!!.moveToFirst()
            val listId = c.getLong(c.getColumnIndex("providerModel"))
            c.close()

            return arrayOf(withId(noteId), fromList(listId), ContentProviderModel.withId(listId))
        }
    }

    @TableEndpoint(name = TestSyncableModel.ENDPOINT, contentProvider = ContentDatabase::class)
    object TestSyncableModel {

        const val ENDPOINT = "TestSyncableModel"

        @ContentUri(path = ENDPOINT, type = ContentUri.ContentType.VND_MULTIPLE + ENDPOINT)
        var CONTENT_URI = buildUri(ENDPOINT)
    }
}
